Documentation/teaching/labs/filesystems_part2.rst

```diff
@@ -0,0 +1,1077 @@
+
+The inode refers to a file on the disk. To refer an open file (associated with a file descriptor within a process), the :c:type:`struct file` structure is used.
+An inode can have any number of (zero or more) ``file`` structures associated (multiple processes can open the same file, or a process can open the same file several times).
+
+Inode exists both as a VFS entity (in memory) and as a disk entity (for UNIX, HFS, NTFS, etc.).
+The inode in VFS is represented by the structure :c:type:`struct inode`.
+Like the other structures in VFS, :c:type:`struct inode` is a generic structure that covers the options for all supported file types, even those that do not have an associated disk entity (such as FAT).
+
+The inode structure
+-------------------
+
+The inode structure is the same for all file systems. In general, file systems also have particular information. These are referenced through the ``i_private`` field of the structure.
+Conventionally, the structure that keeps that particular information is called ``<fsname>_inode_info``, where ``fsname`` represents the file system name. For example, minix and ext4 filesystems store particular information in structures :c:type:`struct minix_inode_info`, or :c:type:`struct ext4_inode_info`.
+
+Some of the important fields of :c:type:`struct inode` are:
+
+  * ``i_sb`` : The superblock structure of the file system to which the inode belongs to.
+  * ``i_rdev``: the device on which this file system is mounted
+  * ``i_ino`` : the number of the inode (uniquely identifies the inode within the file system)
+  * ``i_blkbits``: number of bits used for the block size == log\ :sub:`2`\ (block size)
+  * ``i_mode``, ``i_uid``, ``i_gid``: access rights, uid, gid
+
+  * ``i_size``: file/directory/etc. size in bytes
+  * ``i_mtime``, ``i_atime``, ``i_ctime``: change, access, and creation time
+  * ``i_nlink``: the number of names entries (dentries) that use this inode; for file systems without links (either hard or symbolic) is always set to 1
+  * ``i_blocks``: the number of blocks used by the file (all blocks, not just data); is only used by the quota subsystem
+  * ``i_op``, ``i_fop``: pointers to operations structures: c:type:`struct inode_operations` and :c:type:`struct file_operations`; ``i_mapping->a_ops`` contains a pointer to :c:type:`struct address_space_operations`.
+  * ``i_count``: the inode counter indicating how many kernel components use it.
+
+Some functions that can be used to work with inodes:
+
+  * :c:func:`new_inode`: creates a new inode, sets the ``i_nlink`` field to 1 and initializes ``i_blkbits``, ``i_sb`` and ``i_dev``;
+  * :c:func:`insert_inode_hash`: adds the inode to the hash table of inodes; an interesting effect of this call is that the inode will be written to the disk if it is marked as dirty;
+
+    .. warning::
+
+      An inode created with :c:func:`new_inode` is not in the hash table, and unless you have serious reasons, you must enter it in the hash table;
```

Comment:
This isn't very clear. "unless you have serios resons **not to**, you should add it in the hash table"